File: Bookmark Extension/ShareViewController.m
//
//  ShareViewController.m
//  Bookmark Extension
//
//  Created by Daniel Loewenherz on 8/17/17.
//  Copyright © 2017 Lionheart Software. All rights reserved.
//

@import ASPinboard;
@import LHSCategoryCollection;
@import FMDB;
@import MobileCoreServices;

#import "ShareViewController.h"
#import "PPAddBookmarkViewController.h"
#import "PPPinboardDataSource.h"
#import "PPTheme.h"

@interface ShareViewController ()

@property (nonatomic) BOOL hasToken;

- (void)displayNoURLAlert;

@end

@implementation ShareViewController

- (instancetype)init {
    [PPTheme customizeUIElements];
    self = [super init];
    if (self) {
        self.view.backgroundColor = [UIColor clearColor];
        self.navigationBarHidden = YES;
    }
    return self;
}

- (void)handleInvalidCredentials:(UIViewController *)controller {
    __weak ShareViewController *weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Invalid Token", nil)
                                                                       message:NSLocalizedString(@"Please open Pushpin to refresh your credentials.", nil)
                                                                preferredStyle:UIAlertControllerStyleAlert];
        [alert lhs_addActionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
            if (weakSelf) {
                __strong ShareViewController *strongSelf = weakSelf;

                [strongSelf.extensionContext completeRequestReturningItems:@[] completionHandler:nil];
            }
        }];
        [controller presentViewController:alert animated:YES completion:nil];
    });
}

- (void)presentController:(UINavigationController *)nc token:(NSString *)token {
    PPAddBookmarkViewController *addBookmarkViewController = (PPAddBookmarkViewController *)nc.topViewController;
    addBookmarkViewController.presentingViewControllersExtensionContext = self.extensionContext;
    addBookmarkViewController.tokenOverride = token;

    nc.modalPresentationStyle = UIModalPresentationFormSheet;
    nc.modalTransitionStyle = UIModalTransitionStyleCoverVertical;
    [PPTheme customizeUIElements];

    __weak ShareViewController *weakSelf = self;
    [self presentViewController:nc animated:YES completion:^{
        if (weakSelf) {
            __strong ShareViewController *strongSelf = weakSelf;

            if (strongSelf.hasToken) {
                __weak ShareViewController *_weakSelf = strongSelf;

                dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
                    [[ASPinboard sharedInstance] lastUpdateWithSuccess:^(NSDate *date) {}
                                                               failure:^(NSError *error) {
                        if (_weakSelf) {
                            __strong ShareViewController *_strongSelf = _weakSelf;
                            [_strongSelf handleInvalidCredentials:nc];
                        }
                    }];
                });
            } else {
                [strongSelf handleInvalidCredentials:nc];
            }
        }
    }];
}

- (void)completeWithURLString:(NSString *)urlString
                        title:(NSString *)title
                  description:(NSString *)description
                        token:(NSString *)token {
    if (urlString) {
        // Check if the bookmark is already in the database.
        __block NSDictionary *post = @{@"url": urlString, @"title": title, @"description": description};
        __block NSInteger count;

        NSURL *containerURL = [[NSFileManager defaultManager] containerURLForSecurityApplicationGroupIdentifier:APP_GROUP];
        NSString *path = [containerURL URLByAppendingPathComponent:@"shared.db"].path;
        DLog(@"%@", path);

        __weak ShareViewController *weakSelf = self;
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            [[FMDatabaseQueue databaseQueueWithPath:path] inDatabase:^(FMDatabase *db) {
                FMResultSet *results = [db executeQuery:@"SELECT COUNT(*) AS count, * FROM bookmark WHERE url=?" withArgumentsInArray:@[urlString]];
                [results next];
                count = [results intForColumnIndex:0];

                if (count > 0) {
                    post = [PPPinboardDataSource postFromResultSet:results];
                }

                [results close];
            }];

            if (count == 0) {
                [[ASPinboard sharedInstance] bookmarkWithURL:urlString
                                                     success:^(NSDictionary *post) {
                    NSDictionary *bookmark = @{@"title": post[@"description"],
                                               @"description": post[@"extended"],
                                               @"url": urlString,
                                               @"private": @([post[@"shared"] isEqualToString:@"no"]),
                                               @"unread": @([post[@"toread"] isEqualToString:@"yes"]),
                                               @"tags": post[@"tags"]};

                    dispatch_async(dispatch_get_main_queue(), ^{
                        if (weakSelf) {
                            __strong ShareViewController *strongSelf = weakSelf;
                            UINavigationController *navigation = [PPAddBookmarkViewController addBookmarkViewControllerWithBookmark:bookmark
                                                                                                                             update:@(YES)
                                                                                                                           callback:nil];
                            [strongSelf presentController:navigation token:token];
                        }
                    });
                }
                                                     failure:^(NSError *error) {
                    NSUserDefaults *sharedDefaults = [[NSUserDefaults alloc] initWithSuiteName:APP_GROUP];
                    BOOL readByDefault = [[sharedDefaults objectForKey:@"ReadByDefault"] boolValue];
                    BOOL privateByDefault = [[sharedDefaults objectForKey:@"PrivateByDefault"] boolValue];

                    NSDictionary *bookmark = @{
                        @"title": title,
                        @"url": urlString,
                        @"description": description,
                        @"private": @(privateByDefault),
                        @"unread": @(!readByDefault)
                    };

                    dispatch_async(dispatch_get_main_queue(), ^{
                        if (weakSelf) {
                            __strong ShareViewController *strongSelf = weakSelf;
                            PPNavigationController *navigation = [PPAddBookmarkViewController addBookmarkViewControllerWithBookmark:bookmark
                                                                                                                             update:@(NO)
                                                                                                                           callback:nil];

                            PPAddBookmarkViewController *addBookmarkViewController = (PPAddBookmarkViewController *)navigation.topViewController;

                            if (!title && !description) {
                                [addBookmarkViewController prefillTitleAndForceUpdate:YES];
                            }

                            [strongSelf presentController:navigation token:token];
                        }
                    });
                }];
            } else {
                dispatch_async(dispatch_get_main_queue(), ^{
                    if (weakSelf) {
                        __strong ShareViewController *strongSelf = weakSelf;
                        UINavigationController *navigation = [PPAddBookmarkViewController addBookmarkViewControllerWithBookmark:post
                                                                                                                         update:@(YES)
                                                                                                                       callback:nil];
                        [strongSelf presentController:navigation token:token];
                    }
                });
            }
        });
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    NSUserDefaults *sharedDefaults = [[NSUserDefaults alloc] initWithSuiteName:APP_GROUP];
    NSString *token = [sharedDefaults objectForKey:@"token"];
    DLog(@"%@", APP_GROUP);
    self.hasToken = token.length > 0;

    if (self.hasToken) {
        [[ASPinboard sharedInstance] setToken:token];
    }

    NSExtensionItem *item = self.extensionContext.inputItems.firstObject;
    NSItemProvider *titleItemProvider;
    NSItemProvider *urlItemProvider;
    NSItemProvider *propertyListItemProvider;
    for (NSItemProvider *itemProvider in item.attachments) {
        if ([itemProvider hasItemConformingToTypeIdentifier:(__bridge NSString *)kUTTypeURL]) {
            urlItemProvider = itemProvider;
            continue;
        }

        if ([itemProvider hasItemConformingToTypeIdentifier:(__bridge NSString *)kUTTypePlainText]) {
            titleItemProvider = itemProvider;
            continue;
        }

        if ([itemProvider hasItemConformingToTypeIdentifier:(__bridge NSString *)kUTTypePropertyList]) {
            propertyListItemProvider = itemProvider;
            break;
        }
    }

    __weak ShareViewController *weakSelf = self;
    if (propertyListItemProvider) {
        [propertyListItemProvider loadItemForTypeIdentifier:(__bridge NSString *)kUTTypePropertyList
                                                    options:0
                                          completionHandler:^(NSDictionary *results, NSError *error) {
            NSDictionary *data = results[NSExtensionJavaScriptPreprocessingResultsKey];

            if (weakSelf) {
                __strong ShareViewController *strongSelf = weakSelf;
                [strongSelf completeWithURLString:data[@"url"] title:data[@"title"] description:data[@"selection"] token:token];
            }
        }];
    } else if (titleItemProvider != nil || urlItemProvider != nil) {
        dispatch_group_t group = dispatch_group_create();

        __block NSString *urlString;
        __block NSString *title;

        if (urlItemProvider) {
            dispatch_group_enter(group);
            [urlItemProvider loadItemForTypeIdentifier:(__bridge NSString *)kUTTypeURL
                                               options:0
                                     completionHandler:^(NSURL *url, NSError *error) {
                urlString = url.absoluteString;

                dispatch_group_leave(group);
            }];
        }

        if (titleItemProvider) {
            dispatch_group_enter(group);

            [titleItemProvider loadItemForTypeIdentifier:(__bridge NSString *)kUTTypePlainText
                                                 options:0
                                       completionHandler:^(NSString *text, NSError *error) {
                NSURL *url = [NSURL URLWithString:text];
                if (url) {
                    urlString = text;
                } else {
                    title = text;
                }

                dispatch_group_leave(group);
            }];
        }

        dispatch_group_notify(group, dispatch_get_main_queue(), ^{
            if (title == nil) {
                title = @"";
            }

            if (weakSelf) {
                __strong ShareViewController *strongSelf = weakSelf;

                if (urlString) {
                    [strongSelf completeWithURLString:urlString title:title description:@"" token:token];
                } else {
                    [strongSelf displayNoURLAlert];
                }
            }
        });
    } else {
        [self displayNoURLAlert];
    }
}

- (void)displayNoURLAlert {
    __weak ShareViewController *weakSelf = self;
    UIAlertController *alert = [UIAlertController lhs_alertViewWithTitle:@"No URL Found" message:@"No URL was provided for this webpage. Please try using another browser. If you still experience issues, please contact support."];
    [alert lhs_addActionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
        if (weakSelf) {
            __strong ShareViewController *strongSelf = weakSelf;
            [strongSelf.extensionContext completeRequestReturningItems:@[] completionHandler:nil];
        }
    }];

    [self presentViewController:alert animated:YES completion:nil];
}

@end

File: BookmarkExtension2/ShareViewController.swift
//
//  ShareViewController.swift
//  BookmarkExtension2
//
//  Created by Daniel Loewenherz on 7/10/16.
//  Copyright © 2016 Lionheart Software. All rights reserved.
//

import UIKit
import Social
import ASPinboard
import MobileCoreServices
import FMDB

final class ShareViewController: SLComposeServiceViewController {
    var hasToken = false
    var text: String?
    var url: String?

    override func isContentValid() -> Bool {
        return true
    }

    override func configurationItems() -> [AnyObject]! {
        return []
    }

    func didSelectPost(animated: Bool) {
        super.viewWillAppear(animated)

        let InvalidCredentials: UIViewController -> Void = { controller in
            dispatch_async(dispatch_get_main_queue()) { 
                let alert = UIAlertController(title: NSLocalizedString("Invalid Token", comment: ""), message: NSLocalizedString("Please open Pushpin to refresh your credentials.", comment: ""), preferredStyle: .Alert)
                alert.lhs_addActionWithTitle("OK", style: .Default) { action in
                    controller.dismissViewControllerAnimated(true, completion: { 
                        self.extensionContext?.completeRequestReturningItems([], completionHandler: nil)
                    })
                }

                controller.presentViewController(alert, animated: true, completion: nil)
            }
        }

        // MARK: TODO
        let APP_GROUP = "io.aurora.pushpin"

        guard let defaults = NSUserDefaults.init(suiteName: APP_GROUP) else {
            return
        }

        let token = defaults.stringForKey("token")

        hasToken = token?.characters.count > 0
        text = nil

        if let token = token where hasToken {
            ASPinboard.sharedInstance().token = token
        }

        let PresentController: UINavigationController -> Void = { controller in
            guard let addBookmarkViewController = controller.topViewController as? PPAddBookmarkViewController else {
                return
            }

            addBookmarkViewController.presentingViewControllersExtensionContext = self.extensionContext
            addBookmarkViewController.tokenOverride = token

            controller.modalPresentationStyle = .FormSheet
            controller.modalTransitionStyle = .CoverVertical
            PPTheme.customizeUIElements()

            self.presentViewController(controller, animated: true, completion: {
                if self.hasToken {
                    ASPinboard.sharedInstance().lastUpdateWithSuccess({ _ in }, failure: { error in
                        InvalidCredentials(controller)
                    })
                }
                else {
                    InvalidCredentials(controller)
                }
            })
        }

        let CompletionHandler: (String?, String?, String?) -> Void = { urlString, title, description in
            dispatch_async(dispatch_get_main_queue()) {
                guard let urlString = urlString else {
                    return
                }

                var post: [NSObject: AnyObject] = [
                    "url": urlString
                ]

                if let title = title, let description = description {
                    post["title"] = title
                    post["description"] = description
                }

                var count: Int32 = 0

                guard let containerURL = NSFileManager.defaultManager().containerURLForSecurityApplicationGroupIdentifier(APP_GROUP),
                    let databasePath = containerURL.URLByAppendingPathComponent("shared.db").path else {
                    return
                }

                FMDatabaseQueue(path: databasePath).inDatabase { db in
                    let results = db.executeQuery("SELECT COUNT(*) AS count, * FROM bookmark WHERE url=?", withArgumentsInArray: [urlString])
                    results.next()
                    count = results.intForColumnIndex(0)

                    if count > 0 {
                        post = PPPinboardDataSource.postFromResultSet(results)
                    }
                }

                var navigation: PPNavigationController!
                if count > 0 {
                    navigation = PPAddBookmarkViewController.addBookmarkViewControllerWithBookmark(post, update: true, callback: nil)
                }
                else {
                    let privateByDefault = defaults.boolForKey("PrivateByDefault")
                    let readByDefault = defaults.boolForKey("ReadByDefault")

                    post["private"] = NSNumber(bool: privateByDefault)
                    post["unread"] = NSNumber(bool: !readByDefault)
                    navigation = PPAddBookmarkViewController.addBookmarkViewControllerWithBookmark(post, update: false, callback: nil)

                    if title == nil && description == nil {
                        let addBookmarkController = navigation.topViewController as? PPAddBookmarkViewController
                        addBookmarkController?.prefillTitleAndForceUpdate(true)
                    }
                }

                PresentController(navigation)
            }
        }

        guard let item = extensionContext?.inputItems.first as? NSExtensionItem,
        let attachments = item.attachments as? [NSItemProvider] else {
            return
        }

        for provider in attachments {
            if provider.hasItemConformingToTypeIdentifier(kUTTypeURL as String) {
                provider.loadItemForTypeIdentifier(kUTTypeURL as String, options: nil, completionHandler: { (url, error) in
                    self.url = (url as? NSURL)?.absoluteString
                    CompletionHandler(self.url, self.text, "")
                })
            }

            if provider.hasItemConformingToTypeIdentifier(kUTTypePlainText as String) {
                provider.loadItemForTypeIdentifier(kUTTypePlainText as String, options: nil, completionHandler: { (text, error) in
                    self.text = text as? String
                    CompletionHandler(self.url, self.text, "")
                })
            }

            if provider.hasItemConformingToTypeIdentifier(kUTTypePropertyList as String) {
                provider.loadItemForTypeIdentifier(kUTTypePropertyList as String, options: nil, completionHandler: { (results, error) in
                    guard let results = results as? [String: AnyObject],
                        let data = results[NSExtensionJavaScriptPreprocessingResultsKey] as? [String: AnyObject] else {
                        return
                    }

                    let url = data["url"] as? String
                    let title = data["title"] as? String
                    let description = data["selection"] as? String
                    CompletionHandler(url, title, description)
                })
            }
        }
    }
}
